File: src/shipyard/core/tree_drift.py
"""Working-tree drift detection for `shipyard run` (#249).

Context: `shipyard run` reads the live working tree through every
validation stage (configure → build → test). Edits that land mid-run
— whether from a parallel agent or a human — corrupt the run
silently, then surface as an unrelated-looking compile error 20+
minutes in. #238 shipped a doc-fix + one-liner warning (P3 scope);
this module is the P2 follow-up that fails fast with a specific
error instead.

How it works: compute a content-addressed signature of the working
tree at run start, re-compute at each stage boundary, abort with
``FailureClass.TREE_DRIFT`` when the signature changes. ~milliseconds
per check on a typical repo.

What counts as the tree: git-tracked files with unstaged/staged
modifications, plus all untracked files that aren't gitignored. That's
the set `git ls-files -m -o --exclude-standard` reports. Ignored files
don't count — build outputs and caches routinely change mid-run and
aren't user edits.

Kept in a dedicated module so LocalExecutor's `_run_stages` integration
stays small and both surfaces (executor + future CLI probes) share the
same signature function.
"""

from __future__ import annotations

import hashlib
import subprocess
from pathlib import Path


def compute_signature(cwd: str | Path | None = None) -> str | None:
    """Return a short opaque signature of the working-tree state.

    The signature changes iff any tracked file's on-disk content
    differs, any untracked non-ignored file appears/disappears/changes,
    or HEAD moves. Returns ``None`` when git is unavailable or the
    path isn't a git repo — callers treat that as "can't check, skip
    the guard" so non-git projects aren't penalized.
    """
    cwd_str = str(cwd) if cwd else None

    head = _git_output(["git", "rev-parse", "HEAD"], cwd_str, timeout=5)
    if head is None:
        return None

    # `-m` = tracked files with unstaged modifications.
    # `-o --exclude-standard` = untracked files that aren't gitignored.
    # Staged-but-unmodified files don't show up here but still contribute
    # via the diff hash below — so a `git add` between stages is caught.
    listing = _git_output(
        ["git", "ls-files", "-m", "-o", "--exclude-standard"],
        cwd_str, timeout=30,
    )
    if listing is None:
        return None

    # Also capture the full diff against HEAD so staged changes (which
    # `-m` doesn't list) and content-level edits show up in the hash.
    # `--no-ext-diff` keeps user-configured external diff tools from
    # making the signature non-deterministic.
    diff = _git_output(
        ["git", "diff", "--no-ext-diff", "HEAD"],
        cwd_str, timeout=30,
    )
    if diff is None:
        return None

    h = hashlib.sha256()
    h.update(head.encode())
    h.update(b"\0LS\0")
    h.update(listing.encode())
    h.update(b"\0DIFF\0")
    h.update(diff.encode())

    # Untracked file contents aren't in `git diff HEAD`, so hash them
    # directly. Sort for determinism.
    base = Path(cwd_str) if cwd_str else Path.cwd()
    for path in sorted(p for p in listing.splitlines() if p):
        full = base / path
        h.update(b"\0U\0")
        h.update(path.encode())
        try:
            h.update(full.read_bytes())
        except OSError:
            # Path vanished between the listing and the read (race
            # with a rename/delete). That *is* drift — fold a sentinel
            # into the hash so a subsequent compute sees a different
            # value.
            h.update(b"<missing>")

    return h.hexdigest()[:32]


def list_dirty_paths(cwd: str | Path | None = None) -> list[str]:
    """List paths currently dirty relative to HEAD.

    Shape matches `git status --short` (status prefix + path) so the
    error message surfaces the same view the user would see by running
    git themselves. Empty list when clean or git is unavailable.
    """
    cwd_str = str(cwd) if cwd else None
    output = _git_output(
        ["git", "status", "--short", "--untracked-files=all"],
        cwd_str, timeout=30,
    )
    if output is None:
        return []
    return [line for line in output.splitlines() if line.strip()]


def format_drift_error(
    stage: str,
    initial_paths: list[str],
    current_paths: list[str],
) -> str:
    """Format the user-facing drift error message (#249 acceptance).

    Shows the stage that was about to run and what changed since the
    run started — the initial dirty set vs the current one — so the
    operator / agent knows exactly which files drifted without diffing
    manually.
    """
    lines = [
        f"working tree changed during `shipyard run` (stage={stage}).",
        "mid-run edits produce non-deterministic failures (#238).",
        "re-run after your other edits settle, or use separate "
        "worktrees for parallel work.",
    ]
    initial_set = set(initial_paths)
    current_set = set(current_paths)
    added = sorted(current_set - initial_set)
    removed = sorted(initial_set - current_set)
    if added or removed:
        lines.append("")
        lines.append("what changed:")
        for entry in added:
            lines.append(f"  + {entry}")
        for entry in removed:
            lines.append(f"  - {entry}")
    # Pass the `--allow-tree-drift` hint last so it's visible but not
    # the first thing a reviewer sees — we don't want to advertise the
    # escape hatch over the fix (rebase / separate worktree).
    lines.append("")
    lines.append(
        "(pass --allow-tree-drift to suppress this guard when you "
        "know a build step mutates the tree on purpose.)"
    )
    return "\n".join(lines)


def _git_output(
    cmd: list[str], cwd: str | None, *, timeout: float,
) -> str | None:
    """Run a git command, return stdout on success, None on any failure."""
    try:
        result = subprocess.run(
            cmd, cwd=cwd, capture_output=True, text=True, timeout=timeout,
        )
    except (FileNotFoundError, subprocess.SubprocessError):
        return None
    if result.returncode != 0:
        return None
    return result.stdout

File: src/tree_drift.rs
use std::path::Path;
use std::process::{Command, Stdio};

use sha2::{Digest, Sha256};

/// Short opaque signature for the Git working tree relevant to `shipyard run`.
#[must_use]
pub fn compute_signature(cwd: &Path) -> Option<String> {
    let head = git_output(cwd, &["rev-parse", "HEAD"])?;
    let listing = git_output(cwd, &["ls-files", "-m", "-o", "--exclude-standard"])?;
    let diff = git_output(cwd, &["diff", "--no-ext-diff", "HEAD"])?;

    let mut hasher = Sha256::new();
    hasher.update(head.as_bytes());
    hasher.update(b"\0LS\0");
    hasher.update(listing.as_bytes());
    hasher.update(b"\0DIFF\0");
    hasher.update(diff.as_bytes());

    for path in listing.lines().filter(|line| !line.trim().is_empty()) {
        hasher.update(b"\0U\0");
        hasher.update(path.as_bytes());
        match std::fs::read(cwd.join(path)) {
            Ok(bytes) => hasher.update(bytes),
            Err(_) => hasher.update(b"<missing>"),
        }
    }

    Some(hex::encode(hasher.finalize())[..32].to_owned())
}

/// Return `git status --short` dirty paths, or empty when unavailable.
#[must_use]
pub fn list_dirty_paths(cwd: &Path) -> Vec<String> {
    git_output(cwd, &["status", "--short", "--untracked-files=all"])
        .map(|output| {
            output
                .lines()
                .map(str::trim_end)
                .filter(|line| !line.trim().is_empty())
                .map(ToOwned::to_owned)
                .collect()
        })
        .unwrap_or_default()
}

/// Format the user-facing tree-drift error for a stage boundary.
#[must_use]
pub fn format_drift_error(
    stage: &str,
    initial_paths: &[String],
    current_paths: &[String],
) -> String {
    let mut lines = vec![
        format!("working tree changed during `shipyard run` (stage={stage})."),
        "mid-run edits produce non-deterministic failures (#238).".to_owned(),
        "re-run after your other edits settle, or use separate worktrees for parallel work."
            .to_owned(),
    ];
    let initial = initial_paths
        .iter()
        .map(String::as_str)
        .collect::<std::collections::BTreeSet<_>>();
    let current = current_paths
        .iter()
        .map(String::as_str)
        .collect::<std::collections::BTreeSet<_>>();
    let added = current.difference(&initial).copied().collect::<Vec<_>>();
    let removed = initial.difference(&current).copied().collect::<Vec<_>>();
    if !added.is_empty() || !removed.is_empty() {
        lines.push(String::new());
        lines.push("what changed:".to_owned());
        for entry in added {
            lines.push(format!("  + {entry}"));
        }
        for entry in removed {
            lines.push(format!("  - {entry}"));
        }
    }
    lines.push(String::new());
    lines.push(
        "(pass --allow-tree-drift to suppress this guard when you know a build step mutates the tree on purpose.)"
            .to_owned(),
    );
    lines.join("\n")
}

fn git_output(cwd: &Path, args: &[&str]) -> Option<String> {
    let output = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .output()
        .ok()?;
    output
        .status
        .success()
        .then(|| String::from_utf8_lossy(&output.stdout).to_string())
}

#[cfg(test)]
mod tests {
    use std::process::Command;

    use super::{compute_signature, format_drift_error, list_dirty_paths};

    #[test]
    fn signature_is_stable_when_tree_is_unchanged() {
        let temp = repo();
        assert_eq!(
            compute_signature(temp.path()),
            compute_signature(temp.path())
        );
    }

    #[test]
    fn signature_changes_on_tracked_file_edit() {
        let temp = repo();
        let before = compute_signature(temp.path()).expect("signature");
        std::fs::write(temp.path().join("seed.txt"), "seed\nedit\n").expect("edit");
        let after = compute_signature(temp.path()).expect("signature");
        assert_ne!(before, after);
    }

    #[test]
    fn signature_changes_on_untracked_file_content() {
        let temp = repo();
        std::fs::write(temp.path().join("new.txt"), "v1\n").expect("new");
        let before = compute_signature(temp.path()).expect("signature");
        std::fs::write(temp.path().join("new.txt"), "v2\n").expect("edit");
        let after = compute_signature(temp.path()).expect("signature");
        assert_ne!(before, after);
    }

    #[test]
    fn signature_ignores_gitignored_file() {
        let temp = repo();
        std::fs::write(temp.path().join(".gitignore"), "build/\n").expect("ignore");
        git(temp.path(), &["add", ".gitignore"]);
        git(temp.path(), &["commit", "-q", "-m", "ignore build"]);
        let before = compute_signature(temp.path()).expect("signature");
        std::fs::create_dir_all(temp.path().join("build")).expect("build");
        std::fs::write(temp.path().join("build").join("artifact.o"), "x").expect("artifact");
        let after = compute_signature(temp.path()).expect("signature");
        assert_eq!(before, after);
    }

    #[test]
    fn signature_changes_when_head_moves() {
        let temp = repo();
        let before = compute_signature(temp.path()).expect("signature");
        std::fs::write(temp.path().join("seed.txt"), "seed\nsecond\n").expect("edit");
        git(temp.path(), &["commit", "-aq", "-m", "second"]);
        let after = compute_signature(temp.path()).expect("signature");
        assert_ne!(before, after);
    }

    #[test]
    fn signature_returns_none_outside_git_repo() {
        let temp = tempfile::tempdir().expect("tempdir");
        assert!(compute_signature(temp.path()).is_none());
    }

    #[test]
    fn dirty_paths_match_git_short_shape() {
        let temp = repo();
        std::fs::write(temp.path().join("seed.txt"), "changed\n").expect("edit");
        std::fs::write(temp.path().join("new.txt"), "new\n").expect("new");
        let dirty = list_dirty_paths(temp.path());
        assert!(dirty.iter().any(|line| line.contains("seed.txt")));
        assert!(dirty.iter().any(|line| line.contains("new.txt")));
    }

    #[test]
    fn drift_error_names_stage_changed_paths_and_escape_hatch() {
        let initial = vec![" M already-dirty.cpp".to_owned()];
        let current = vec![
            " M already-dirty.cpp".to_owned(),
            "?? new-change.cpp".to_owned(),
        ];
        let message = format_drift_error("build", &initial, &current);
        assert!(message.contains("stage=build"));
        assert!(message.contains("#238"));
        assert!(message.contains("+ ?? new-change.cpp"));
        assert!(!message.contains("+  M already-dirty.cpp"));
        assert!(message.contains("--allow-tree-drift"));
    }

    fn repo() -> tempfile::TempDir {
        let temp = tempfile::tempdir().expect("tempdir");
        git(temp.path(), &["init", "--quiet", "--initial-branch=main"]);
        std::fs::write(temp.path().join("seed.txt"), "seed\n").expect("seed");
        git(temp.path(), &["add", "."]);
        git(temp.path(), &["commit", "-q", "-m", "seed"]);
        temp
    }

    fn git(cwd: &std::path::Path, args: &[&str]) {
        let status = Command::new("git")
            .args(args)
            .current_dir(cwd)
            .env("GIT_AUTHOR_NAME", "T")
            .env("GIT_AUTHOR_EMAIL", "t@t")
            .env("GIT_COMMITTER_NAME", "T")
            .env("GIT_COMMITTER_EMAIL", "t@t")
            .status()
            .expect("git runs");
        assert!(status.success(), "git failed: {args:?}");
    }
}
